packages/babel-plugin-transform-class-properties/src/index.js

```diff
@@ -9,21 +9,212 @@ export default function({ types: t }) {
+      if (computed || !isStatic) return;
+
+      const name = t.isIdentifier(key) ? key.name : key.value;
+      if (name === "prototype") {
+        throw path.buildCodeFrameError("illegal static class field");
       }
     },
+
+    PrivateName(path) {
+      const { parentPath, node } = path;
+      if (parentPath.isMemberExpression({ property: node, computed: false })) {
+        if (!this.privateProps[node.id.name]) {
+          throw path.buildCodeFrameError("unknown private property");
+        }
+
+        return;
+      }
+
+      if (parentPath.isClassPrivateProperty({ key: node })) return;
+
+      throw path.buildCodeFrameError(
+        `illegal syntax. Did you mean \`this.#${node.id.name}\`?`,
+      );
+    },
+  };
+
+  const privateNameRemapper = {
+    PrivateName(path) {
+      const { node, parent, parentPath, scope } = path;
+      if (node.id.name !== this.name) {
+        return;
+      }
+      if (!parentPath.isMemberExpression()) return;
+
+      const grandParentPath = parentPath.parentPath;
+      const { object } = parent;
+
+      let replacePath = parentPath;
+      let replaceWith = t.callExpression(this.get, [object, this.privateMap]);
+
+      if (
+        grandParentPath.isAssignmentExpression() ||
+        grandParentPath.isUpdateExpression()
+      ) {
+        const { node } = grandParentPath;
+        let assign;
+        let memo;
+        let postfix;
+
+        if (grandParentPath.isAssignmentExpression({ operator: "=" })) {
+          assign = node.right;
+        } else {
+          const { right, operator } = node;
+          memo = scope.maybeGenerateMemoised(object);
+
+          if (memo) {
+            replaceWith.arguments[0] = memo;
+            memo = t.assignmentExpression("=", memo, object);
+          }
+
+          if (grandParentPath.isUpdateExpression({ prefix: false })) {
+            postfix = scope.generateUidIdentifierBasedOnNode(parent);
+            scope.push({ id: postfix });
+            replaceWith = t.assignmentExpression(
+              "=",
+              postfix,
+              t.unaryExpression("+", replaceWith),
+            );
+          }
+
+          assign = t.binaryExpression(
+            operator.slice(0, -1),
+            replaceWith,
+            right || t.numericLiteral(1),
+          );
+        }
+
+        replacePath = grandParentPath;
+        replaceWith = t.callExpression(this.put, [
+          memo || object,
+          this.privateMap,
+          assign,
+        ]);
+
+        if (postfix) {
+          replaceWith = t.sequenceExpression([replaceWith, postfix]);
+        }
+      } else if (grandParentPath.isCallExpression({ callee: parent })) {
+        const memo = scope.maybeGenerateMemoised(object);
+        if (memo) {
+          replaceWith.arguments[0] = t.assignmentExpression("=", memo, object);
+        }
+
+        const call = t.clone(grandParentPath.node);
+        call.callee = t.memberExpression(replaceWith, t.identifier("call"));
```

Comment:
Should there be any attempt to call this function with the original Function.prototype.call, rather than whatever the `call` property is of that function? (I'm not sure what Babel does here in general.)